crates/amaru-sim/src/bin/amaru-sim/sync.rs

```diff
@@ -0,0 +1,194 @@
+use amaru_kernel::Point;
+use amaru_ouroboros::protocol::{peer::Peer, PullEvent};
+use gasket::framework::*;
+use maelstrom::protocol::Message;
+use serde::{Deserialize, Serialize};
+use std::io;
+use tracing::{error, trace_span, Span};
+
+use crate::bytes::Bytes;
+
+pub type DownstreamPort = gasket::messaging::OutputPort<PullEvent>;
+
+pub enum WorkUnit {
+    Send(PullEvent),
+}
+
+#[derive(Stage)]
+#[stage(name = "pull", unit = "WorkUnit", worker = "Worker")]
+pub struct Stage {
+    pub downstream: DownstreamPort,
+}
+
+impl Stage {
+    pub(crate) fn new(_tip: &Point) -> Self {
+        Self {
+            downstream: DownstreamPort::default(),
+        }
+    }
+}
+
+pub struct Worker {}
+
+#[async_trait::async_trait(?Send)]
+impl gasket::framework::Worker<Stage> for Worker {
+    async fn bootstrap(_stage: &Stage) -> Result<Self, WorkerError> {
+        let worker = Self {};
+
+        Ok(worker)
+    }
+
+    async fn schedule(
+        &mut self,
+        _stage: &mut Stage,
+    ) -> Result<WorkSchedule<WorkUnit>, WorkerError> {
+        // read one line of stdin which should be a JSON-formatted message from
+        // some peer to our peer
+        let mut input = String::new();
+        let span = trace_span!("pull-worker");
+
+        io::stdin()
+            .read_line(&mut input)
+            .map_err(|_| WorkerError::Recv)?;
+
+        match serde_json::from_str::<Message>(input.as_str()) {
```

Comment:
Why use serde_json here ? Or, more specifically, why JSON as a transport given that it's mostly CBOR bytes being transferred. Wouldn't it make more sense to take a CBOR object entirely and decode it?

Reply:
It's the maelstrom interface. We could change it but then this would become annoying down the road because Jepsen does not speak CBOR. We could of course change Jepsen but then just using JSON here seems more portable and simpler.

Reply:
(Maelstrom's [protocol](https://github.com/jepsen-io/maelstrom/blob/main/doc/protocol.md) uses JSON. Jepsen proper can use any codec (as long as you can write it in Clojure/Java). In the simulation-testing repo the codec is a parameter.)